controllers/harborconfiguration_controller.go

```diff
@@ -0,0 +1,421 @@
+			SrcRegistry:   srcRegistry,
+			DestNamespace: harborConfiguration.Spec.Replication.DestinationNamespace,
+			DestRegistry:  reqDestinationRegistry,
+			Filters:       reqFilters,
+			Trigger:       reqTrigger,
+			Override:      harborConfiguration.Spec.Replication.Override,
+			Enabled:       harborConfiguration.Spec.Replication.EnablePolicy,
+		}
+
+		err = client.UpdateReplicationPolicy(ctx, &update, replicationFound.ID)
+		// Experienced some flakiness
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *HarborConfigurationReconciler) reconcileAll(ctx context.Context, harborConfiguration harborconfigurationv1alpha1.HarborConfiguration, client *apiv2.RESTClient) (ctrl.Result, error) {
+	registry := &modelv2.Registry{
+		Name:        harborConfiguration.Spec.Registry.Name,
+		Type:        harborConfiguration.Spec.Registry.Type,
+		URL:         harborConfiguration.Spec.Registry.TargetRegistryUrl,
+		Description: harborConfiguration.Spec.Registry.Description,
+		Credential:  (*modelv2.RegistryCredential)(harborConfiguration.Spec.Registry.Credential),
+	}
+
+	_, err := r.registryReconciliation(ctx, harborConfiguration, *registry, client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	_, err = r.projectReconciliation(ctx, harborConfiguration, *registry, client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	_, err = r.replicationRuleReconciliation(ctx, harborConfiguration, *registry, client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, err
+}
+
+func getHarborSecret(ctx context.Context, clientSet *kubernetes.Clientset, harborcluster *harborOperator.HarborCluster) (string, error) {
+	passwordSecret, err := clientSet.CoreV1().Secrets(harborcluster.Namespace).Get(ctx, harborcluster.Spec.HarborAdminPasswordRef, v1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	return string(passwordSecret.Data["secret"]), nil
+}
+
+func getHarborURL(harborcluster *harborOperator.HarborCluster) string {
+	url := fmt.Sprintf("http://%s-harbor-harbor-core.%s/api/v2.0", harborcluster.Name, harborcluster.Namespace)
+	return url
+}
+
+func getTypedKubeConfig() (*kubernetes.Clientset, error) {
+	var config *rest.Config
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		config, err = getkubeConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	kubeConfig, err := kubernetes.NewForConfig(config)
+
+	return kubeConfig, err
+}
+
+func getDynamicKubeConfig() (dynamic.Interface, error) {
+	var config *rest.Config
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		config, err = getkubeConfig()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		fmt.Printf("error creating dynamic client: %v\n", err)
+		os.Exit(1)
+	}
+
+	return dynamicClient, err
+}
+
+func getkubeConfig() (*rest.Config, error) {
+	userHomeDir, err := os.UserHomeDir()
```

Comment:
`client-go` has a pretty extensive [util.HomeDir()](https://pkg.go.dev/k8s.io/client-go@v0.25.4/util/homedir#HomeDir) function for this. It's useful, because it covers more bases (Windows, Darwin).